[TTI][TLI] Support scalable immediates with isLegalAddImmediate

huntergr-arm opened the pull request:
Adds a second parameter (default to 0) to isLegalAddImmediate, to represent a scalable immediate.

Extends the AArch64 implementation to match immediates based on vscale * base_register_size * immediate in the range [-32,31].

See the vscale-aware LSR RFC for reference: https://discourse.llvm.org/t/rfc-vscale-aware-loopstrengthreduce/77131

huntergr-arm committed:
[AArch64] Recognize legal add immediates for addvl

davemgreen commented:
Can you remind us where these functions are used? It might be simpler to have separate methods and move the scalable vs normal immediate check out a level.

RKSimon reviewed (commented):
I'm confused as to how this will work - how do you indicate when to test for a legal scalar or scalable immediate? Suppose you want to know if 0 is legal?

huntergr-arm committed:
[TTI][TLI][AArch64] Add isLegalAddScalableImmediate

Adds an interface to determine whether an immediate would be legal within
an add instruction, when said immediate is multiplied by vscale.

davemgreen commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -696,6 +696,12 @@ class TargetTransformInfo {
   /// immediate without having to materialize the immediate into a register.
   bool isLegalAddImmediate(int64_t Imm) const;
 
+  /// Return true if the specified immediate is legal add of a scalable
+  /// immediate, that is the target has add instructions which can add a
+  /// register with the immediate (multiplied by vscale) without having to
+  /// materialize the immediate into a register.
```
Maybe:
"Return true if the add of the specified scalable immediate is legal, that is..."

davemgreen commented on llvm/unittests/Target/AArch64/Immediates.cpp:
```diff
@@ -0,0 +1,62 @@
+    // +(16 * vscale), one register's worth
+    {16, true},
+    // +(8 * vscale), half a register's worth
+    {8, false},
```
8 could be a `inch` if that wouldn't cause problems with your analysis. Same for 2 and 4.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -16595,6 +16595,21 @@ bool AArch64TargetLowering::isLegalAddImmediate(int64_t Immed) const {
 
+bool AArch64TargetLowering::isLegalAddScalableImmediate(int64_t Imm) const {
+  // Scalable immediates require SVE support.
+  if (!Subtarget->hasSVE())
```
Should all of this require sve2? Otherwise we don't generate the instructions.

davemgreen reviewed (commented):
This is looking sensible to me, I have some comments below from experimenting with some values.

huntergr-arm committed:
Require SVE2, support inc/dec immediates, adjust doxygen comment

davemgreen reviewed (approved):
Thanks. This looks sensible to me.
